File: services/api/app/models/community.py
from typing import Optional
from pydantic import Field

from app.models.domain import CamelModel


class CreatePostRequest(CamelModel):
    title: str
    body: str
    course_code: Optional[str] = None
    professor_name: Optional[str] = None
    is_anonymous: bool = False
    general_tags: list[str] = Field(default_factory=list)


class CreateReplyRequest(CamelModel):
    body: str
    parent_reply_id: Optional[str] = None
    is_anonymous: bool = False


class PostSummary(CamelModel):
    id: str
    user_id: str
    title: str
    body: str
    course_code: Optional[str] = None
    professor_name: Optional[str] = None
    is_anonymous: bool = False
    general_tags: list[str] = Field(default_factory=list)
    author_display_name: str
    created_at: str
    updated_at: str
    reply_count: int = 0
    upvote_count: int = 0
    downvote_count: int = 0
    user_has_upvoted: bool = False
    user_has_downvoted: bool = False


class ReplyOut(CamelModel):
    id: str
    post_id: str
    user_id: str
    body: str
    parent_reply_id: Optional[str] = None
    is_anonymous: bool = False
    author_display_name: str
    created_at: str
    updated_at: str
    upvote_count: int = 0
    downvote_count: int = 0
    user_has_upvoted: bool = False
    user_has_downvoted: bool = False


class PostDetail(PostSummary):
    replies: list[ReplyOut] = Field(default_factory=list)


class PostListResponse(CamelModel):
    posts: list[PostSummary]
    total: int
    page: int
    page_size: int


class UpvoteResponse(CamelModel):
    """Kept for backward compatibility."""
    upvoted: bool
    upvote_count: int


class VoteResponse(CamelModel):
    voted: bool
    upvote_count: int
    downvote_count: int


class NotificationOut(CamelModel):
    id: str
    user_id: str
    type: str
    payload: Optional[dict] = None
    read: bool
    created_at: str

File: frontend/src/types/community.ts
export interface PostSummary {
  id: string;
  userId: string;
  title: string;
  body: string;
  courseCode: string | null;
  professorName: string | null;
  isAnonymous: boolean;
  generalTags: string[];
  authorDisplayName: string;
  createdAt: string;
  updatedAt: string;
  replyCount: number;
  upvoteCount: number;
  downvoteCount: number;
  userHasUpvoted: boolean;
  userHasDownvoted: boolean;
}

export interface ReplyOut {
  id: string;
  postId: string;
  userId: string;
  body: string;
  parentReplyId: string | null;
  isAnonymous: boolean;
  authorDisplayName: string;
  createdAt: string;
  updatedAt: string;
  upvoteCount: number;
  downvoteCount: number;
  userHasUpvoted: boolean;
  userHasDownvoted: boolean;
}

export interface PostDetail extends PostSummary {
  replies: ReplyOut[];
}

export interface PostListResponse {
  posts: PostSummary[];
  total: number;
  page: number;
  pageSize: number;
}

export interface CreatePostPayload {
  title: string;
  body: string;
  courseCode?: string;
  professorName?: string;
  isAnonymous?: boolean;
  generalTags?: string[];
}

export interface CreateReplyPayload {
  body: string;
  parentReplyId?: string;
  isAnonymous?: boolean;
}

export interface UpvoteResponse {
  upvoted: boolean;
  upvoteCount: number;
}

export interface VoteResponse {
  voted: boolean;
  upvoteCount: number;
  downvoteCount: number;
}

export interface NotificationOut {
  id: string;
  userId: string;
  type: string;
  payload: Record<string, unknown> | null;
  read: boolean;
  createdAt: string;
}

export type SortBy = "newest" | "best";
